packages/connection-manager/src/js/ConnectionManager/ConnectionManager.js

```diff
@@ -0,0 +1,102 @@
+import { List } from "immutable";
+// import { AuthStore } from "authenticator";
+import ConnectionQueue from "../ConnectionQueue/ConnectionQueue.js";
+import { INIT, STARTED } from "../Connection/StateConstants";
+
+/**
+ * The Connection Manager which is responsible for
+ * queuing Connections into the ConnectionQueue and
+ * actually starting them, when they are head of
+ * waiting line.
+ */
+export default class ConnectionManager {
+  /**
+   * Initializes an Instance of ConnectionManager
+   * @param {Integer} maxConnections – max open connections
+   */
+  constructor(maxConnections = 6) {
+    Object.defineProperty(this, "maxConnections", {
+      get() {
+        return maxConnections;
+      }
+    });
+
+    let waiting = new ConnectionQueue();
+    Object.defineProperty(this, "waiting", {
+      get() {
+        return waiting;
+      },
+      set(_waiting) {
+        waiting = _waiting;
+      }
+    });
+
+    let open = List();
+    Object.defineProperty(this, "open", {
+      get() {
+        return open;
+      },
+      set(_open) {
+        open = _open;
+      }
+    });
+
+    this.handleConnectionClose = this.handleConnectionClose.bind(this);
+  }
+
+  /**
+   * Queues given connection with given priority
+   * @param {AbstractConnection} connection – connection to queue
+   * @param {Integer} [priority=0] – optional change of priority, will be updated inside connection
+   */
+  queue(connection, priority = 0) {
+    if (
+      (connection.state === INIT || connection.state === STARTED) &&
+      !connection.listeners("close").includes(this.handleConnectionClose)
+    ) {
+      connection.on("close", this.handleConnectionClose);
+    }
+    switch (connection.state) {
+      case INIT:
+        this.waiting = this.waiting.enqueue(connection, priority);
+        break;
+      case STARTED:
+        this.open = this.open.add(connection);
+        break;
+    }
+
+    this.next();
+  }
+  /**
+   * handles all queue activity events
+   * @param {AbstractConnection} connection
+   * @return {void}
+   */
+  next() {
+    if (this.open.size >= this.maxConnections || this.waiting.size === 0) {
+      return;
+    }
+
+    const connection = this.waiting.head();
+    this.waiting = this.waiting.tail();
```

Comment:
Let's revise `this.waiting`. The methods we're using on this `Queue` are a little strange. A Queue has `enqueue`, `dequeue`, and a few other helpers for peeking and checking size/empty. I don't understand why we've introduced `head` and `tail` methods. A `head` and  `tail` are usually reserved for internal use by the queue to track position and depend on the `queue`s implementation. Because we use a `List` as the core storage mechanism, I don't see why we even need to track head/tail since a `List` in Javascript provides out of the box methods for this.

Reply:
I have added this to JIRA for further discussion, but we should leave it as it is now for this PR.

Reply:
The respective functions _mirror_ the immutable [Queue](http://www.scala-lang.org/api/2.12.0/scala/collection/immutable/Queue.html) (part of Scala standard lib) implementation and follow a well known pattern thus I don't think we need to revise them. But we might want to rethink the general naming as one could argue that's not a queue due to the sorting we're applying which violates the FIFO principal.

Reply:
The function names sparked quite some discussion and thus even though they follow well known patterns we're going revisit the names and try to align them with non immutable implementations. /cc @dcos/frontend-foundation

Reply:
@juliangieseke I'm sorry to bring up more PITA things and appreciate your patience and diligence on this. The immutable aspect is what makes this tricky. My suggestion would be to keep `tail()` as long as we rename `head()` to `peek()`.

Reply:
@MatApple Having a `peek` and `tail` method sounds wrong to me as that would really follow any pattern. We're going to either drop immutable.js and rewrite the queue to provide mutable operations similar to those found in Java or expose the immutable.js interface (exposing `first` and `shift` should be sufficient) instead of following any known implementation. /cc @dcos/frontend-foundation @juliangieseke

Reply:
changed `head` to `first` and `tail` to `shift`. 👍